sc-poc/modules/pubsub/websocket.go

```diff
@@ -0,0 +1,165 @@
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gobwas/ws"
+	"github.com/gobwas/ws/wsutil"
+	"github.com/google/uuid"
+	"github.com/mitchellh/mapstructure"
+	"go.uber.org/zap"
+
+	"github.com/spacecloud-io/space-cloud/managers/apis"
+)
+
+// TODO: channels may or may not have prefix slash
+
+// GetRoutes returns all the apis that are exposed by this app
+func (a *App) GetAPIRoutes() apis.APIs {
+	channels := a.Channels()
+	apis := []*apis.API{}
+
+	for path, channel := range channels.Channels {
+		// Get the publish and subscribe API of the channel
+		publisherAPI := a.getPublisherAPI(path, channel.Name)
+		subscriberAPI := a.getSubscriberAPI(path, channel.Name)
+
+		apis = append(apis, publisherAPI, subscriberAPI)
+	}
+
+	return apis
+}
+
+// getPublishAPI creates a websocket API for sending messages in the channel
+func (a *App) getPublisherAPI(channelPath, channelName string) *apis.API {
+	return &apis.API{
+		Name: fmt.Sprintf("%s-publisher", channelName),
+		Path: fmt.Sprintf("/v1/pubsub/default%s/publisher", channelPath),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Create the websocket connection
+			conn, _, _, err := ws.UpgradeHTTP(r, w)
+			if err != nil {
+				a.logger.Error("could not establish websocket connection", zap.String("channel", channelName), zap.Error(err))
+				return
+			}
+
+			// Go routine to receive messages from the websocket connection and publish
+			// them into the watermill channel
+			go func() {
```

Comment:
I don't think we need to run this as a seperate goroutine.

You can confirm this by starting a subscriber then trying to reload the config. If the config gets reloaded without affecting the socket we are good.